### PR TITLE
making pbr work

### PR DESCRIPTION
**Pull request by bonjorno7**
oh boy
the lookup table was wrong, I replaced it with an approximate function that actually works
direct lighting wasn't working because of a weird condition in the for loop
implemented ambient occlusion, multiplies over the combined lighting
implemented emission texture, adds to the combined lighting

### CODE REVIEW

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 1]
```diff
@@ -250,10 +248,10 @@ END_SHADER_PARAMS
 
 			if (bHasMraoTexture)
 				BindTexture(SHADER_SAMPLER10, info.mraoTexture, 0);
-			else
-				pShaderAPI->BindStandardTexture(SHADER_SAMPLER10, TEXTURE_WHITE);
+	  else
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: Would you mind using tab indents here as well?

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 2]
```diff
@@ -251,41 +277,39 @@ float4 main( PS_INPUT i ) : COLOR
...
+	for (uint n = 0; n < NUM_LIGHTS; ++n) // now lighting works on one model but not on another??
 	{
 		// Li
 		float3 LightIn = normalize(PixelShaderGetLightVector(i.worldPos, cLightInfo, n));
 		float3 LightColor = PixelShaderGetLightColor(cLightInfo, n) * GetAttenForLight(i.lightAtten, n);
 
 		directLighting += calculateLight(LightIn, LightColor, outgoingLightDirection,
 			normal, fresnelReflectance, roughness, metalness, lightDirectionAngle, albedo);
+
+		//return float4(directLighting.x + directLighting.y + directLighting.z, 0.0, 0.5, 1.0); // testing if the loop runs at all
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Do we still need this?

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 3]
```diff
@@ -251,41 +277,39 @@ float4 main( PS_INPUT i ) : COLOR
 
 	float3 directLighting = 0.0;
 
-	for (uint n = 0; n < NUM_LIGHTS && n < 0; ++n)
+	for (uint n = 0; n < NUM_LIGHTS; ++n) // now lighting works on one model but not on another??
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Is this due to the flashlight thing?

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 4]
```diff
@@ -321,7 +345,9 @@ float4 main( PS_INPUT i ) : COLOR
 
 	float fogFactor = CalcPixelFogFactor( PIXELFOGTYPE, g_FogParams, g_EyePos.z, i.worldPos.z, i.projPos.z );
 
-	float alpha = 1.0;
+	//float alpha = 1.0;
+	float alpha = albedo.a; // Gocnak told me to add this
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: This is ok, can you remove the comment above and to the right?

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 5]
```diff
@@ -333,5 +359,11 @@ float4 main( PS_INPUT i ) : COLOR
 	//return float4(g_LightmapCoords[0].x + i.baseTexCoord.x * g_LightmapCoords[0].z, g_LightmapCoords[0].z, i.baseTexCoord.x, 1);
 	//return tex2D(LightmapSampler, g_LightmapCoords[0].xy + i.baseTexCoord * g_LightmapCoords[0].z);
 
-	return FinalOutput( float4( directLighting + ambientLighting, alpha ), fogFactor, PIXELFOGTYPE, TONEMAP_SCALE_LINEAR, bWriteDepthToAlpha, i.projPos.z );	
+#if EMISSION
+	float3 combinedLighting = (directLighting + ambientLighting) * ambientOcclusion + emission;
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Would you mind setting `combinedLighting` above the `#if` and doing `combinedLighting += emission;` inside the `#if`?

**Review by thexa4 — CHANGES_REQUESTED** [review]

**Comment by ksharperd** [comment]
That good,you finally repair it!

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp (hunk 1 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: done

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 2 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: nope, removed now

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 4 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: done

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 5 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: alright

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 3 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: nope, it's the same with or without the flashlight on, I just can't get direct lighting to work consistently
perhaps I can fix that in the next update
at least it's better than it was :/

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 6]
```diff
@@ -25,6 +26,7 @@
 
 #include "common_flashlight_fxc.h"
 #include "common_lightmappedgeneric_fxc.h"
+#include "common_vertexlitgeneric_dx9.h"
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Why do we need this extra include?

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 7]
```diff
@@ -33,16 +35,14 @@ const float3 cAmbientCube[6]				: register( PSREG_AMBIENT_CUBE );
...
 const float4x4 g_FlashlightWorldToTexture	: register( PSREG_FLASHLIGHT_TO_WORLD_TEXTURE );
 PixelShaderLightInfo cLightInfo[3]			: register( PSREG_LIGHT_INFO_ARRAY );				// 2 registers each - 6 registers total (4th light spread across w's)
-//const float4 g_LightmapCoords[3]			: register( PSREG_AMBIENT_CUBE );
 const float4 g_BaseColor					: register( PSREG_SELFILLUMTINT );
 
-#define PI								3.141592
-#define EPSILON							0.00001
+static const float PI = 3.141592;
+static const float EPSILON = 0.00001;
+static const float ENVMAPLOD = 6.0;
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Could you add a comment explaining what the 6.0 stands for? (1 roughness == 4x4 texture, 0 roughness == 256x256 texture)

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 8]
```diff
@@ -154,7 +156,7 @@ float3 calculateLight(float3 lightIn, float3 lightIntensity, float3 lightOut, fl
 	float G = gaSchlickGGX(cosLightIn, lightDirectionAngle, roughness);
 
 	// Diffuse scattering happens due to light being refracted multiple times by a dielectric medium.
-	// Metals on the other hand either reflect or absorb energy, so diffuse contribution is always zero.
+	// Metals on the other hand either reflect or absorb energso diffuse contribution is alwaysy,  zero.
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: could you revert or fix these changes?

**Review by thexa4 — CHANGES_REQUESTED** [review]

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 9]
```diff
@@ -321,17 +330,19 @@ float4 main( PS_INPUT i ) : COLOR
...
 
 	bool bWriteDepthToAlpha = ( WRITE_DEPTH_TO_DESTALPHA != 0 ) && ( WRITEWATERFOGTODESTALPHA == 0 );
-	//ambientLighting = 0;
-	
-	//return i.lightmapTexCoord;
-	//return float4(g_LightmapCoords[0].x + i.baseTexCoord.x * g_LightmapCoords[0].z, g_LightmapCoords[0].z, i.baseTexCoord.x, 1);
-	//return tex2D(LightmapSampler, g_LightmapCoords[0].xy + i.baseTexCoord * g_LightmapCoords[0].z);
 
-	return FinalOutput( float4( directLighting + ambientLighting, alpha ), fogFactor, PIXELFOGTYPE, TONEMAP_SCALE_LINEAR, bWriteDepthToAlpha, i.projPos.z );	
+	float3 combinedLighting = (directLighting + ambientLighting) * ambientOcclusion;
```
Comment (thexa4) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Ambient occlusion should only apply to `ambientLighting`, not to both.

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 6 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: we don't, yeet

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 7 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: // 6 is the lowest cubemap mipmap we need. in a 256x256 cubemap, it gives us 4x4

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 8 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: dunno how that happened, fixed

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 9 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: yep

**Pull request merged**